### DISCUSSION TITLE
Error on "brew uninstall --cask wireshark" / "brew install --cask wireshark": the package path specified was invalid

### DISCUSSION THREAD

**Discussion by Danrancan** [Homebrew/discussions#4190] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.21-38-gc9496ae
ORIGIN: https://github.com/Homebrew/brew
HEAD: c9496ae50fefd506b1d75882d3283edc518bdf20
Last commit: 2 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 9a3bb2f2da346d28fa6c431d835ef4c31f9f86c9
Core tap last commit: 2 weeks ago
Core tap branch: master
Core tap JSON: 09 Feb 17:20 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.4arRxMZ1fU/org.xquartz:0
HOMEBREW_EDITOR: nano
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 13.0.0 build 1300
Git: 2.32.0 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.7.3-x86_64
CLT: 13.2.0.0.1.1638488800
Xcode: 13.2.1
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I cannot uninstall, reinstall, or remove the wireshark cask with homebrew.
```
$brew uninstall --cask wireshark
==> Uninstalling Cask wireshark
Error: Failure while executing; `/usr/bin/sudo -E -- /usr/sbin/installer -pkg /usr/local/Caskroom/wireshark/4.0.2/Uninstall\ ChmodBPF.pkg -target /` exited with 1. Here's the output:
installer: Error - the package path specified was invalid: '/usr/local/Caskroom/wireshark/4.0.2/Uninstall ChmodBPF.pkg'.
```
```
$ brew reinstall --cask wireshark
==> Downloading https://2.na.dl.wireshark.org/osx/Wireshark%204.0.3%20Intel%2064.dmg
Already downloaded: /Users/danran/Library/Caches/Homebrew/downloads/a6f03fe549a414f3714d9e9e2820119a2476a9e772ad7640c6c31055ca19b3af--Wireshark 4.0.3 Intel 64.dmg
==> Uninstalling Cask wireshark
Error: Failure while executing; `/usr/bin/sudo -E -- /usr/sbin/installer -pkg /usr/local/Caskroom/wireshark/4.0.2/Uninstall\ ChmodBPF.pkg -target /` exited with 1. Here's the output:
installer: Error - the package path specified was invalid: '/usr/local/Caskroom/wireshark/4.0.2/Uninstall ChmodBPF.pkg'.
```
```
$ brew uninstall --cask wireshark
==> Uninstalling Cask wireshark
Error: Failure while executing; `/usr/bin/sudo -E -- /usr/sbin/installer -pkg /usr/local/Caskroom/wireshark/4.0.2/Uninstall\ ChmodBPF.pkg -target /` exited with 1. Here's the output:
installer: Error - the package path specified was invalid: '/usr/local/Caskroom/wireshark/4.0.2/Uninstall ChmodBPF.pkg'.
```

Could someone please help? I just want a fully working fully upgradable wireshark cask. I have no idea how to fix this.

**Comment by cho-m** [comment]
Given it is complaining about missing `/usr/local/Caskroom/wireshark/4.0.2/Uninstall ChmodBPF.pkg`, can you check the path?

Maybe see what is inside directory:
```
ls /usr/local/Caskroom/wireshark/4.0.2
```

The Cask uninstall logic hasn't changed in last few releases. Could be some incomplete uninstall but hard to tell without additional info.

Final resort would be manual uninstall and/or `--force` arg.

---

I at least tried out installing/uninstalling current 4.0.3 version and didn't see an issue there.

**Reply by Danrancan** [reply]
> ls /usr/local/Caskroom/wireshark/4.0.2

```
ls /usr/local/Caskroom/wireshark/4.0.2
ls: cannot access '/usr/local/Caskroom/wireshark/4.0.2': No such file or directory
```

I guess the directory doesn't exist. I'm thinking I had a partial install/uninstall that was happening when the computer turned off mid-install. But not sure what to do from here. How do I use the --force arg?

`brew uninstall --force wireshark` renders the exact same errors.

Nevermind. I figured it out. `brew install --force wireshark` works. Then `brew uninstall wireshark` also works. 
Same goes for `brew install --cask --force wireshark` and `brew uninstall --cask wireshark`

**Reply by ThomasMinor** [reply]
That was a good hint but I had to remove the installation directory manually first to run `brew --force` successfully.

**Comment by adrienthebo** [comment]
I'm running into this as well; I wound up installing the wireshark CLI first and unintentionally aborted installation of the cask in the middle of installation.

It looks like the "Uninstall ChmodBPF.pkg" postinstall script is brittle:

```shell
set -e

CHMOD_BPF_PLIST="/Library/LaunchDaemons/org.wireshark.ChmodBPF.plist"
BPF_GROUP="access_bpf"

launchctl unload -F "$CHMOD_BPF_PLIST"

dscl . -read /Groups/"$BPF_GROUP" > /dev/null 2>&1 && \
    dseditgroup -q -o delete "$BPF_GROUP"

rm -rf "/Library/Application Support/Wireshark"

rm -f "$CHMOD_BPF_PLIST"

# ChmodBPF hasn't been a startup item since 2018 (ac4f3c0f4d).
rm -rf /Library/StartupItems/ChmodBPF

# https://gitlab.com/wireshark/wireshark/-/issues/18734
pkgutil --forget org.wireshark.ChmodBPF.pkg
```

Since we're running `set -e` the absence of the launchd service or group will cause the script to fail. I've made some incremental progress in stubbing out the required files so the uninstaller can complete but have to come back to this topic, but in short I suspect that stubbing out the dependencies that the uninstaller scripts depend on might work around this limitation.

**Comment by Ender-Wang** [comment]
I went into the exact same problem recently, brew version is 4.0.7, upcoming update is 4.0.8, and have tried uninstalling with brew, and the same output as you have posted. I found a workaround: I downloaded 4.0.8 ARM from wireshark.org, installed it, and then went back to the terminal to run `brew upgrade -g`, and it got upgraded from 4.0.7 to 4.0.8.

**Comment by say4n** [comment]
What worked for me was `rm -rf /opt/homebrew/Caskroom/wireshark/` followed by a `brew install --cask wireshark`!

**Reply by vlad88sv** [reply]
I also had to `rm -rf /Applications/Wireshark.app`

**Comment by rofrol** [comment]
```shell
#!/usr/bin/env bash
VERSION="$1"
if [ -z "${VERSION}" ]; then
	echo "Please set VERSION"
	exit 1
fi
# fd pkg /opt/homebrew/Caskroom/wireshark/${VERSION}/Wireshark.app
ln -s /opt/homebrew/Caskroom/wireshark/${VERSION}/Wireshark.app/Contents/Resources/Extras/Remove\ Wireshark\ from\ the\ system\ path.pkg /opt/homebrew/Caskroom/wireshark/${VERSION}/
ln -s /opt/homebrew/Caskroom/wireshark/${VERSION}/Wireshark.app/Contents/Resources/Extras/Uninstall\ ChmodBPF.pkg /opt/homebrew/Caskroom/wireshark/${VERSION}/
brew reinstall --cask wireshark

```

and run:

`sh brew_reinstall_wireshark.sh 4.2.2`

**Comment by bdruth** [comment]
This worked for me:

`rm -rf /opt/homebrew/Caskroom/wireshark/ && rm -rf /Applications/Wireshark.app && ls -l /opt/homebrew/bin/* | grep -i wireshark | awk '{print $9}' | xargs rm -rf && brew install --cask wireshark`
